webapp/src/serial.tsx

```diff
@@ -0,0 +1,371 @@
+                        lastEntry.insertBefore(newLabel, lastEntry.lastChild)
+                    }
+                } else {
+                    //make a new non-collapsed entry
+                    newEntry.appendChild(document.createTextNode(this.consoleBuffer))
+                    this.consoleRoot.appendChild(newEntry)
+                    this.consoleRoot.scrollTop = this.consoleRoot.scrollHeight
+                }
+                if (this.consoleRoot.childElementCount > this.maxConsoleEntries) {
+                    this.consoleRoot.removeChild(this.consoleRoot.firstChild)
+                }
+                this.consoleBuffer = ""
+            }
+        }
+    }
+
+    pauseRecording() {
+        this.active = false
+        if (this.recordIcon) this.recordIcon.className = "circle icon"
+        if (this.recordButton) {
+            this.recordButton.classList.remove("green")
+            this.recordButton.classList.add("circular")
+            this.recordButton.classList.add("red")
+        }
+        this.charts.forEach(s => s.stop())
+    }
+
+    startRecording() {
+        this.active = true
+        if (this.recordIcon) this.recordIcon.className = "pause icon"
+        if (this.recordButton) {
+            this.recordButton.classList.remove("red")
+            this.recordButton.classList.remove("circular")
+            this.recordButton.classList.add("green")
+        }
+        this.charts.forEach(s => s.start())
+    }
+
+    toggleRecording() {
+        pxt.tickEvent("serial.toggleRecording")
+        if (this.active) this.pauseRecording()
+        else this.startRecording()
+    }
+
+    clearNode(e: HTMLElement) {
+        while (e.hasChildNodes()) {
+            e.removeChild(e.firstChild)
+        }
+    }
+
+    clear() {
+        //TODO use refs
+        let chartRoot = document.getElementById("serialCharts")
+        let consoleRoot = document.getElementById("serialConsole")
+        this.clearNode(chartRoot)
+        this.clearNode(consoleRoot)
+        this.charts = []
+        this.consoleBuffer = ""
+    }
+
+    entriesToPlaintext() {
+        return this.rawDataBuffer
+    }
+
+    showExportDialog() {
+        pxt.tickEvent("serial.showExportDialog")
+        const targetTheme = pxt.appTarget.appTheme
+        let rootUrl = targetTheme.embedUrl
+        if (!rootUrl) {
+            pxt.commands.browserDownloadAsync(this.entriesToPlaintext(), "data.txt", "text/plain")
+            return
+        }
+        if (!/\/$/.test(rootUrl)) rootUrl += '/'
+
+        core.confirmAsync({
+            logos: undefined,
+            header: lf("Analyze Data"),
+            hideAgree: true,
+            disagreeLbl: lf("Close"),
+            onLoaded: (_) => {
+                _.find('#datasavelocalfile').click(() => {
+                    pxt.tickEvent("serial.dataExported")
+                    _.modal('hide')
+                    pxt.commands.browserDownloadAsync(this.entriesToPlaintext(), "data.txt", "text/plain")
```

Comment:
is CSV output coming back?

Reply:
Post-merge, yes.

Reply:
See https://github.com/Microsoft/pxt/issues/3120 and https://github.com/Microsoft/pxt/issues/3118